clang/lib/AST/Decl.cpp

```diff
@@ -4159,7 +4159,7 @@ FunctionTemplateDecl *FunctionDecl::getPrimaryTemplate() const {
   if (FunctionTemplateSpecializationInfo *Info
         = TemplateOrSpecialization
             .dyn_cast<FunctionTemplateSpecializationInfo*>()) {
-    return Info->getTemplate();
+    return Info->getTemplate()->getMostRecentDecl();
```

Comment:
Why do we need to get the most recent declaration in all these places?

I worry this will cause us problems in the future, where for example in friend declarations, the exact redeclaration can be significant.

Won't this also degrade source representation accuracy?

Reply:
Most recent guarantees you have 'the most' information about the declaration, but there might be value in getting the definition if available, else most recent?

Reply:
> Why do we need to get the most recent declaration in all these places?

@mizvekov One of the changes made by this patch is to set the `isMemberSpecialization` flag on a per-declaration basis. This ensures that `getTemplateInstantiationArgs` behaves correctly for implicitly instantiated declarations that are explicitly specialized for an implicitly instantiated specialization of their enclosing class template.

When we look up the name of a template, we typically find the most recent declaration of that template. If that declaration is an explicit specialization of a member template, we use the correct declaration when we try to find the definition of that template. However, if the member template is explicitly specialized _after_ we lookup its name, we end up with an expression naming the implicitly instantiated in-class declaration of that template. If we call `getTemplateInstantiationPattern` on that declaration, it will incorrectly return the pattern from the primary template.

Reply:
Changing the implementation of getPrimaryTemplate has a global effect, while you are interested in fixing a behavior specific to `getTemplateInstantiationPattern`. Why can't we fix this behavior there instead?

I am worried about two things:
* The most recent declaration is not always the *right* declaration you have to follow in order to get to the pattern.
* We will degrade source fidelity for code currently using these helpers, as they won't be able to find the exact redeclaration for their primary template anymore.

Reply:
@mizvekov I reverted the change where `getPrimaryTemplate` returns the most recent declaration & changed `getTemplateInstantiationPattern` to use the most recent declaration of the primary template instead in d312bd464bae21d127f468075cb78178f782ca9a.

Reply:
CC @zyn0217 & @erichkeane

Reply:
> @mizvekov I reverted the change where `getPrimaryTemplate` returns the most recent declaration & changed `getTemplateInstantiationPattern` to use the most recent declaration of the primary template instead in [d312bd4](https://github.com/llvm/llvm-project/commit/d312bd464bae21d127f468075cb78178f782ca9a).

That seems reasonable to me.

Reply:
Thanks!